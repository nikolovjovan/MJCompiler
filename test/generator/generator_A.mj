// level A code generation:
// - assignment, variable increment, variable decrement designator statements
// - designator statement, read and print statements
// - expressions, terms, all factors except method call w/ arguments (only w/o)
// - array access designator
// - global/local variables (int, char, bool or array of int, char or bool)
// - global constants
program level_a_code_generation
    const int N = 25, M = 15;
    const char SPACE = ' ', C = 'C';
    const bool T = true, F = false;

    int x, y, z[];
    char c;
    bool ind;
{
    void printHello() {
        print('H'); print('e'); print('l'); print('l'); print('o');
        print(SPACE);
        print('W'); print('o'); print('r'); print('l'); print('d'); print('!');
        print(eol);
    }

    int getValue() {
        return (x - y) * N - z[N - 1] % M;
    }

    void main()
        int a, b, arr[];
        char msg[];
        bool valid;
    {
        // method call
        printHello();

        // global constants print
        print('N'); print(N); print(SPACE); print('M'); print(M); print(eol);
        print('C'); print(C); print(eol);
        print('T'); print(T); print(SPACE); print('F'); print(F); print(eol);

        // global variables
        x = 5;
        print('x'); print(x); print(eol);
        y = x += 2;
        y++;
        x--;
        print('x'); print(x); print(SPACE); print('y'); print(y); print(eol);
        z = new int[N];
        z[x] = -5;
        print('z'); print('['); print(x); print(']'); print(z[x]); print(eol);
        c = chr(65);
        print('c'); print(c); print(eol);
        ind = T;
        print('i'); print('n'); print('d'); print(ind); print(eol);

        // local variables
        a = z[x] *= -x + N * M % (M - y);
        print('z'); print('['); print(x); print(']'); print(z[x]); print(SPACE); print('a'); print(a); print(eol);
        b = ord('C');
        print('b'); print(b); print(SPACE);
        b += (N - 2) * (M + 5) + (y / 2);
        print('b'); print(b); print(eol);
        arr = new int[(M * 2) + 7];
        print('l'); print('e'); print('n'); print(len(arr)); print(eol);
        msg = new char[8];
        msg[0] = 'M'; msg[1] = 'e'; msg[2] = 's'; msg[3] = 's'; msg[4] = 'a'; msg[5] = 'g'; msg[6] = 'e'; msg[7] = eol;
        print(msg[0]); print(msg[1]); print(msg[2]); print(msg[3]); print(msg[4]); print(msg[5]); print(msg[6]); print(msg[7]);
        valid = true;
        print('v'); print('a'); print('l'); print('i'); print('d'); print(valid); print(eol);

        // read global variables
        read(x);
        print('x'); print(x); print(eol);
        read(z[N - 1]);
        print('z'); print('['); print(N - 1); print(']'); print(z[N - 1]); print(eol);
        read(c);
        print('c'); print(c); print(eol);
        read(ind);
        print('i'); print('n'); print('d'); print(ind); print(eol);

        // read local variables
        read(a);
        print('a'); print(a); print(eol);
        read(msg[0]);
        print('m'); print('s'); print('g'); print('['); print(0); print(']'); print(msg[0]); print(eol);
        read(valid);
        print('v'); print('a'); print('l'); print('i'); print('d'); print(valid); print(eol);

        print(chr(65)); print(ord('C')); print(eol);

        print('a'); print(a); print(SPACE);
        a = getValue();
        print('a'); print(a); print(eol);
    }
}