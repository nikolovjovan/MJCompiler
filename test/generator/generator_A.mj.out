Compiling source file: D:\Documents\Fakultet\7. semestar\13E114PP1\Projekat\MJCompiler\test\generator\generator_A.mj
Syntax analysis completed without errors!
Starting semantic analysis...
Found symbolic constant usage 'SPACE' at line 19! Symbol node: Constant 'SPACE': type = 'char', adr = 32, level = 0
Found symbolic constant usage 'eol' at line 21! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found global variable usage 'x' at line 25! Symbol node: Variable 'x': type = 'int', adr = 0, level = 0
Found global variable usage 'y' at line 25! Symbol node: Variable 'y': type = 'int', adr = 1, level = 0
Found symbolic constant usage 'N' at line 25! Symbol node: Constant 'N': type = 'int', adr = 25, level = 0
Found global variable usage 'z' at line 25! Symbol node: Variable 'z': type = 'array of int', adr = 2, level = 0
Found symbolic constant usage 'N' at line 25! Symbol node: Constant 'N': type = 'int', adr = 25, level = 0
Found array element access 'N[N - 1]' at line 25! Symbol node: Array element 'N[N - 1]': type = 'int', adr = -1, level = -1
Found symbolic constant usage 'M' at line 25! Symbol node: Constant 'M': type = 'int', adr = 15, level = 0
Found global method call 'printHello' at line 34! Symbol node: Method 'printHello': type = 'void', adr = 0, level = 0
Found symbolic constant usage 'N' at line 37! Symbol node: Constant 'N': type = 'int', adr = 25, level = 0
Found symbolic constant usage 'SPACE' at line 37! Symbol node: Constant 'SPACE': type = 'char', adr = 32, level = 0
Found symbolic constant usage 'M' at line 37! Symbol node: Constant 'M': type = 'int', adr = 15, level = 0
Found symbolic constant usage 'eol' at line 37! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found symbolic constant usage 'C' at line 38! Symbol node: Constant 'C': type = 'char', adr = 67, level = 0
Found symbolic constant usage 'eol' at line 38! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found symbolic constant usage 'T' at line 39! Symbol node: Constant 'T': type = 'bool', adr = 1, level = 0
Found symbolic constant usage 'SPACE' at line 39! Symbol node: Constant 'SPACE': type = 'char', adr = 32, level = 0
Found symbolic constant usage 'F' at line 39! Symbol node: Constant 'F': type = 'bool', adr = 0, level = 0
Found symbolic constant usage 'eol' at line 39! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found global variable usage 'x' at line 42! Symbol node: Variable 'x': type = 'int', adr = 0, level = 0
Found global variable usage 'x' at line 43! Symbol node: Variable 'x': type = 'int', adr = 0, level = 0
Found symbolic constant usage 'eol' at line 43! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found global variable usage 'y' at line 44! Symbol node: Variable 'y': type = 'int', adr = 1, level = 0
Found global variable usage 'x' at line 44! Symbol node: Variable 'x': type = 'int', adr = 0, level = 0
Found global variable usage 'y' at line 45! Symbol node: Variable 'y': type = 'int', adr = 1, level = 0
Found global variable usage 'x' at line 46! Symbol node: Variable 'x': type = 'int', adr = 0, level = 0
Found global variable usage 'x' at line 47! Symbol node: Variable 'x': type = 'int', adr = 0, level = 0
Found symbolic constant usage 'SPACE' at line 47! Symbol node: Constant 'SPACE': type = 'char', adr = 32, level = 0
Found global variable usage 'y' at line 47! Symbol node: Variable 'y': type = 'int', adr = 1, level = 0
Found symbolic constant usage 'eol' at line 47! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found global variable usage 'z' at line 48! Symbol node: Variable 'z': type = 'array of int', adr = 2, level = 0
Found symbolic constant usage 'N' at line 48! Symbol node: Constant 'N': type = 'int', adr = 25, level = 0
Found global variable usage 'z' at line 49! Symbol node: Variable 'z': type = 'array of int', adr = 2, level = 0
Found global variable usage 'x' at line 49! Symbol node: Variable 'x': type = 'int', adr = 0, level = 0
Found array element access 'x[x]' at line 49! Symbol node: Array element 'x[x]': type = 'int', adr = -1, level = -1
Found global variable usage 'x' at line 50! Symbol node: Variable 'x': type = 'int', adr = 0, level = 0
Found global variable usage 'z' at line 50! Symbol node: Variable 'z': type = 'array of int', adr = 2, level = 0
Found global variable usage 'x' at line 50! Symbol node: Variable 'x': type = 'int', adr = 0, level = 0
Found array element access 'x[x]' at line 50! Symbol node: Array element 'x[x]': type = 'int', adr = -1, level = -1
Found symbolic constant usage 'eol' at line 50! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found global variable usage 'c' at line 51! Symbol node: Variable 'c': type = 'char', adr = 3, level = 0
Found global method call 'chr' at line 51! Symbol node: Method 'chr': type = 'char', adr = 0, level = 1
  Formal parameter 'i': type = 'int', adr = 0, level = 1
Found global variable usage 'c' at line 52! Symbol node: Variable 'c': type = 'char', adr = 3, level = 0
Found symbolic constant usage 'eol' at line 52! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found global variable usage 'ind' at line 53! Symbol node: Variable 'ind': type = 'bool', adr = 4, level = 0
Found symbolic constant usage 'T' at line 53! Symbol node: Constant 'T': type = 'bool', adr = 1, level = 0
Found global variable usage 'ind' at line 54! Symbol node: Variable 'ind': type = 'bool', adr = 4, level = 0
Found symbolic constant usage 'eol' at line 54! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found local variable usage 'a' at line 57! Symbol node: Variable 'a': type = 'int', adr = 0, level = 1
Found global variable usage 'z' at line 57! Symbol node: Variable 'z': type = 'array of int', adr = 2, level = 0
Found global variable usage 'x' at line 57! Symbol node: Variable 'x': type = 'int', adr = 0, level = 0
Found array element access 'x[x]' at line 57! Symbol node: Array element 'x[x]': type = 'int', adr = -1, level = -1
Found global variable usage 'x' at line 57! Symbol node: Variable 'x': type = 'int', adr = 0, level = 0
Found symbolic constant usage 'N' at line 57! Symbol node: Constant 'N': type = 'int', adr = 25, level = 0
Found symbolic constant usage 'M' at line 57! Symbol node: Constant 'M': type = 'int', adr = 15, level = 0
Found symbolic constant usage 'M' at line 57! Symbol node: Constant 'M': type = 'int', adr = 15, level = 0
Found global variable usage 'y' at line 57! Symbol node: Variable 'y': type = 'int', adr = 1, level = 0
Found global variable usage 'x' at line 58! Symbol node: Variable 'x': type = 'int', adr = 0, level = 0
Found global variable usage 'z' at line 58! Symbol node: Variable 'z': type = 'array of int', adr = 2, level = 0
Found global variable usage 'x' at line 58! Symbol node: Variable 'x': type = 'int', adr = 0, level = 0
Found array element access 'x[x]' at line 58! Symbol node: Array element 'x[x]': type = 'int', adr = -1, level = -1
Found symbolic constant usage 'SPACE' at line 58! Symbol node: Constant 'SPACE': type = 'char', adr = 32, level = 0
Found local variable usage 'a' at line 58! Symbol node: Variable 'a': type = 'int', adr = 0, level = 1
Found symbolic constant usage 'eol' at line 58! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found local variable usage 'b' at line 59! Symbol node: Variable 'b': type = 'int', adr = 1, level = 1
Found global method call 'ord' at line 59! Symbol node: Method 'ord': type = 'int', adr = 0, level = 1
  Formal parameter 'ch': type = 'char', adr = 0, level = 1
Found local variable usage 'b' at line 60! Symbol node: Variable 'b': type = 'int', adr = 1, level = 1
Found symbolic constant usage 'SPACE' at line 60! Symbol node: Constant 'SPACE': type = 'char', adr = 32, level = 0
Found local variable usage 'b' at line 61! Symbol node: Variable 'b': type = 'int', adr = 1, level = 1
Found symbolic constant usage 'N' at line 61! Symbol node: Constant 'N': type = 'int', adr = 25, level = 0
Found symbolic constant usage 'M' at line 61! Symbol node: Constant 'M': type = 'int', adr = 15, level = 0
Found global variable usage 'y' at line 61! Symbol node: Variable 'y': type = 'int', adr = 1, level = 0
Found local variable usage 'b' at line 62! Symbol node: Variable 'b': type = 'int', adr = 1, level = 1
Found symbolic constant usage 'eol' at line 62! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found local variable usage 'arr' at line 63! Symbol node: Variable 'arr': type = 'array of int', adr = 2, level = 1
Found symbolic constant usage 'M' at line 63! Symbol node: Constant 'M': type = 'int', adr = 15, level = 0
Found global method call 'len' at line 64! Symbol node: Method 'len': type = 'int', adr = 0, level = 1
  Formal parameter 'arr': type = 'array of none', adr = 0, level = 1
Found local variable usage 'arr' at line 64! Symbol node: Variable 'arr': type = 'array of int', adr = 2, level = 1
Found symbolic constant usage 'eol' at line 64! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found local variable usage 'msg' at line 65! Symbol node: Variable 'msg': type = 'array of char', adr = 3, level = 1
Found local variable usage 'msg' at line 66! Symbol node: Variable 'msg': type = 'array of char', adr = 3, level = 1
Found array element access 'msg[0]' at line 66! Symbol node: Array element 'msg[0]': type = 'char', adr = -1, level = -1
Found local variable usage 'msg' at line 66! Symbol node: Variable 'msg': type = 'array of char', adr = 3, level = 1
Found array element access 'msg[1]' at line 66! Symbol node: Array element 'msg[1]': type = 'char', adr = -1, level = -1
Found local variable usage 'msg' at line 66! Symbol node: Variable 'msg': type = 'array of char', adr = 3, level = 1
Found array element access 'msg[2]' at line 66! Symbol node: Array element 'msg[2]': type = 'char', adr = -1, level = -1
Found local variable usage 'msg' at line 66! Symbol node: Variable 'msg': type = 'array of char', adr = 3, level = 1
Found array element access 'msg[3]' at line 66! Symbol node: Array element 'msg[3]': type = 'char', adr = -1, level = -1
Found local variable usage 'msg' at line 66! Symbol node: Variable 'msg': type = 'array of char', adr = 3, level = 1
Found array element access 'msg[4]' at line 66! Symbol node: Array element 'msg[4]': type = 'char', adr = -1, level = -1
Found local variable usage 'msg' at line 66! Symbol node: Variable 'msg': type = 'array of char', adr = 3, level = 1
Found array element access 'msg[5]' at line 66! Symbol node: Array element 'msg[5]': type = 'char', adr = -1, level = -1
Found local variable usage 'msg' at line 66! Symbol node: Variable 'msg': type = 'array of char', adr = 3, level = 1
Found array element access 'msg[6]' at line 66! Symbol node: Array element 'msg[6]': type = 'char', adr = -1, level = -1
Found local variable usage 'msg' at line 66! Symbol node: Variable 'msg': type = 'array of char', adr = 3, level = 1
Found array element access 'msg[7]' at line 66! Symbol node: Array element 'msg[7]': type = 'char', adr = -1, level = -1
Found symbolic constant usage 'eol' at line 66! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found local variable usage 'msg' at line 67! Symbol node: Variable 'msg': type = 'array of char', adr = 3, level = 1
Found array element access 'msg[0]' at line 67! Symbol node: Array element 'msg[0]': type = 'char', adr = -1, level = -1
Found local variable usage 'msg' at line 67! Symbol node: Variable 'msg': type = 'array of char', adr = 3, level = 1
Found array element access 'msg[1]' at line 67! Symbol node: Array element 'msg[1]': type = 'char', adr = -1, level = -1
Found local variable usage 'msg' at line 67! Symbol node: Variable 'msg': type = 'array of char', adr = 3, level = 1
Found array element access 'msg[2]' at line 67! Symbol node: Array element 'msg[2]': type = 'char', adr = -1, level = -1
Found local variable usage 'msg' at line 67! Symbol node: Variable 'msg': type = 'array of char', adr = 3, level = 1
Found array element access 'msg[3]' at line 67! Symbol node: Array element 'msg[3]': type = 'char', adr = -1, level = -1
Found local variable usage 'msg' at line 67! Symbol node: Variable 'msg': type = 'array of char', adr = 3, level = 1
Found array element access 'msg[4]' at line 67! Symbol node: Array element 'msg[4]': type = 'char', adr = -1, level = -1
Found local variable usage 'msg' at line 67! Symbol node: Variable 'msg': type = 'array of char', adr = 3, level = 1
Found array element access 'msg[5]' at line 67! Symbol node: Array element 'msg[5]': type = 'char', adr = -1, level = -1
Found local variable usage 'msg' at line 67! Symbol node: Variable 'msg': type = 'array of char', adr = 3, level = 1
Found array element access 'msg[6]' at line 67! Symbol node: Array element 'msg[6]': type = 'char', adr = -1, level = -1
Found local variable usage 'msg' at line 67! Symbol node: Variable 'msg': type = 'array of char', adr = 3, level = 1
Found array element access 'msg[7]' at line 67! Symbol node: Array element 'msg[7]': type = 'char', adr = -1, level = -1
Found local variable usage 'valid' at line 68! Symbol node: Variable 'valid': type = 'bool', adr = 4, level = 1
Found local variable usage 'valid' at line 69! Symbol node: Variable 'valid': type = 'bool', adr = 4, level = 1
Found symbolic constant usage 'eol' at line 69! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found global variable usage 'x' at line 72! Symbol node: Variable 'x': type = 'int', adr = 0, level = 0
Found global variable usage 'x' at line 73! Symbol node: Variable 'x': type = 'int', adr = 0, level = 0
Found symbolic constant usage 'eol' at line 73! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found global variable usage 'z' at line 74! Symbol node: Variable 'z': type = 'array of int', adr = 2, level = 0
Found symbolic constant usage 'N' at line 74! Symbol node: Constant 'N': type = 'int', adr = 25, level = 0
Found array element access 'N[N - 1]' at line 74! Symbol node: Array element 'N[N - 1]': type = 'int', adr = -1, level = -1
Found symbolic constant usage 'N' at line 75! Symbol node: Constant 'N': type = 'int', adr = 25, level = 0
Found global variable usage 'z' at line 75! Symbol node: Variable 'z': type = 'array of int', adr = 2, level = 0
Found symbolic constant usage 'N' at line 75! Symbol node: Constant 'N': type = 'int', adr = 25, level = 0
Found array element access 'N[N - 1]' at line 75! Symbol node: Array element 'N[N - 1]': type = 'int', adr = -1, level = -1
Found symbolic constant usage 'eol' at line 75! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found global variable usage 'c' at line 76! Symbol node: Variable 'c': type = 'char', adr = 3, level = 0
Found global variable usage 'c' at line 77! Symbol node: Variable 'c': type = 'char', adr = 3, level = 0
Found symbolic constant usage 'eol' at line 77! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found global variable usage 'ind' at line 78! Symbol node: Variable 'ind': type = 'bool', adr = 4, level = 0
Found global variable usage 'ind' at line 79! Symbol node: Variable 'ind': type = 'bool', adr = 4, level = 0
Found symbolic constant usage 'eol' at line 79! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found local variable usage 'a' at line 82! Symbol node: Variable 'a': type = 'int', adr = 0, level = 1
Found local variable usage 'a' at line 83! Symbol node: Variable 'a': type = 'int', adr = 0, level = 1
Found symbolic constant usage 'eol' at line 83! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found local variable usage 'msg' at line 84! Symbol node: Variable 'msg': type = 'array of char', adr = 3, level = 1
Found array element access 'msg[0]' at line 84! Symbol node: Array element 'msg[0]': type = 'char', adr = -1, level = -1
Found local variable usage 'msg' at line 85! Symbol node: Variable 'msg': type = 'array of char', adr = 3, level = 1
Found array element access 'msg[0]' at line 85! Symbol node: Array element 'msg[0]': type = 'char', adr = -1, level = -1
Found symbolic constant usage 'eol' at line 85! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found local variable usage 'valid' at line 86! Symbol node: Variable 'valid': type = 'bool', adr = 4, level = 1
Found local variable usage 'valid' at line 87! Symbol node: Variable 'valid': type = 'bool', adr = 4, level = 1
Found symbolic constant usage 'eol' at line 87! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found global method call 'chr' at line 89! Symbol node: Method 'chr': type = 'char', adr = 0, level = 1
  Formal parameter 'i': type = 'int', adr = 0, level = 1
Found global method call 'ord' at line 89! Symbol node: Method 'ord': type = 'int', adr = 0, level = 1
  Formal parameter 'ch': type = 'char', adr = 0, level = 1
Found symbolic constant usage 'eol' at line 89! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Found local variable usage 'a' at line 91! Symbol node: Variable 'a': type = 'int', adr = 0, level = 1
Found symbolic constant usage 'SPACE' at line 91! Symbol node: Constant 'SPACE': type = 'char', adr = 32, level = 0
Found local variable usage 'a' at line 92! Symbol node: Variable 'a': type = 'int', adr = 0, level = 1
Found global method call 'getValue' at line 92! Symbol node: Method 'getValue': type = 'int', adr = 0, level = 0
Found local variable usage 'a' at line 93! Symbol node: Variable 'a': type = 'int', adr = 0, level = 1
Found symbolic constant usage 'eol' at line 93! Symbol node: Constant 'eol': type = 'char', adr = 10, level = 0
Semantic analysis completed without errors!
====================================================================================================
    6 symbolic constant(s) defined
    5 global variable(s) defined
    0 class(es) defined
    3 global method(s) defined
======================================== SYMBOL TABLE DUMP =========================================
Type 'int': type = 'int', adr = -1, level = -1
Type 'char': type = 'char', adr = -1, level = -1
Type 'bool': type = 'bool', adr = -1, level = -1
Constant 'null': type = 'class', adr = 0, level = 0
Constant 'eol': type = 'char', adr = 10, level = 0
Method 'chr': type = 'char', adr = 0, level = 1
  Formal parameter 'i': type = 'int', adr = 0, level = 1
Method 'ord': type = 'int', adr = 0, level = 1
  Formal parameter 'ch': type = 'char', adr = 0, level = 1
Method 'len': type = 'int', adr = 0, level = 1
  Formal parameter 'arr': type = 'array of none', adr = 0, level = 1
Program 'level_a_code_generation': type = 'none', adr = 0, level = 1
  Constant 'N': type = 'int', adr = 25, level = 0
  Constant 'M': type = 'int', adr = 15, level = 0
  Constant 'SPACE': type = 'char', adr = 32, level = 0
  Constant 'C': type = 'char', adr = 67, level = 0
  Constant 'T': type = 'bool', adr = 1, level = 0
  Constant 'F': type = 'bool', adr = 0, level = 0
  Variable 'x': type = 'int', adr = 0, level = 0
  Variable 'y': type = 'int', adr = 1, level = 0
  Variable 'z': type = 'array of int', adr = 2, level = 0
  Variable 'c': type = 'char', adr = 3, level = 0
  Variable 'ind': type = 'bool', adr = 4, level = 0
  Method 'printHello': type = 'void', adr = 0, level = 0
  Method 'getValue': type = 'int', adr = 0, level = 0
  Method 'main': type = 'void', adr = 0, level = 0
    Variable 'a': type = 'int', adr = 0, level = 1
    Variable 'b': type = 'int', adr = 1, level = 1
    Variable 'arr': type = 'array of int', adr = 2, level = 1
    Variable 'msg': type = 'array of char', adr = 3, level = 1
    Variable 'valid': type = 'bool', adr = 4, level = 1

Generating bytecode file: D:\Documents\Fakultet\7. semestar\13E114PP1\Projekat\MJCompiler\test\generator_A.obj
Compilation finished!
